pkg/virt-api/rest/subresource.go

```diff
@@ -420,37 +510,40 @@ func (app *SubresourceAPIApp) StopVMRequestHandler(request *restful.Request, res
 	response.WriteHeader(http.StatusAccepted)
 }
 
-func (app *SubresourceAPIApp) findPod(namespace string, vmi *v1.VirtualMachineInstance) (string, error) {
-	fieldSelector := fields.ParseSelectorOrDie("status.phase==" + string(k8sv1.PodRunning))
-	labelSelector, err := labels.Parse(fmt.Sprintf(v1.AppLabel + "=virt-launcher," + v1.CreatedByLabel + "=" + string(vmi.UID)))
-	if err != nil {
-		return "", err
+func (app *SubresourceAPIApp) findVirtHandlerPodForVMI(vmi *v1.VirtualMachineInstance) (*k8sv1.Pod, error) {
+	if !vmi.IsRunning() {
+		return nil, goerror.New(fmt.Sprintf("VirtualMachineInstance's phase is %s instead of %s", vmi.Status.Phase, v1.Running))
 	}
-	selector := k8smetav1.ListOptions{FieldSelector: fieldSelector.String(), LabelSelector: labelSelector.String()}
 
-	podList, err := app.VirtCli.CoreV1().Pods(namespace).List(selector)
+	fieldSelector := fields.ParseSelectorOrDie("spec.nodeName==" + vmi.Status.NodeName)
+	labelSelector, err := labels.Parse(fmt.Sprintf(v1.AppLabel + "=virt-handler"))
 	if err != nil {
-		return "", err
+		return nil, err
 	}
+	selector := k8smetav1.ListOptions{FieldSelector: fieldSelector.String(), LabelSelector: labelSelector.String()}
 
-	if len(podList.Items) == 0 {
-		return "", goerror.New("connection failed. No VirtualMachineInstance pod is running")
-	} else if len(podList.Items) == 1 {
-		return podList.Items[0].ObjectMeta.Name, nil
-	} else {
-		// If we have 2 running pods, we might have a migration. Find the new pod!
-		if vmi.Status.MigrationState != nil && vmi.Status.MigrationState.Completed {
-			for _, pod := range podList.Items {
-				if pod.Name == vmi.Status.MigrationState.TargetPod {
-					return pod.Name, nil
-				}
+	podList, err := app.VirtCli.CoreV1().Pods(namespaceKubevirt).List(selector)
+	if err != nil {
+		return nil, err
+	}
+
+	switch {
+	case len(podList.Items) == 1:
```

Comment:
nit: you can remove first case because you return it under the default and IMHO `if` conditions more suitable here than `switch`(not critical :smile:)

Reply:
I agree that it's not so significant in this particular case because the evaluation of all case statements is trivial but as I expect the first case to hold in more than 90% of the cases, I prefer not to evaluate the other case statements in those cases (who knows, maybe the migration case will require a query in the future or we'll add another complex case..)